### PR TITLE
fix a MaterialApp NNBD issue

### LINKED ISSUE #81504: Stateless widget without home results in null check error
Running Flutter cli on chrome OS - linux buster 

Status:
==========================================
:: flutter upgrade
Flutter is already up to date on channel stable
Flutter 2.0.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision adc687823a (13 days ago) • 2021-04-16 09:40:20 -0700
Engine • revision b09f014e96
Tools • Dart 2.12.3
:: flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.0.5, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[✓] Android Studio
[✓] VS Code (version 1.55.2)
[✓] Connected device (3 available)

• No issues found!
:: flutter devices
3 connected devices:

Google Pixelbook (mobile) • emulator-5554 • android-x64    • Android 9 (API 28)
Linux (desktop)           • linux         • linux-x64      • Linux
Chrome (web)              • chrome        • web-javascript • Chromium 90.0.4430.85
built on Debian 10.9, running on Debian 10.9
========================================

Command:
=====================================
flutter run -d emulator-5554
=====================================

main.dart
====================================
import 'package:flutter/material.dart';
void main() => runApp(Myapp());
  
class Myapp extends StatelessWidget 
{
  @override
  Widget build(BuildContext context) 
  {
    return MaterialApp
    (
      title: 'Flutter Demo',
      theme: ThemeData
      (
        primarySwatch: Colors.blue,
      ),
    );
  }
}
======================================

compiles without error. 
produces unexpected application display
and execution error messages. 

see screenshot below for image of execution:

Screenshot 2021-04-29 16 08 06

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #81504 
Fix crashing when set `MaterialApp.home` to null.
 Also, fix a crashing when update `MaterialApp.home` to non-null from null.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/app.dart [hunk 1]
```diff
@@ -1170,7 +1170,7 @@ class _WidgetsAppState extends State<WidgetsApp> with WidgetsBindingObserver {
       assert(!_usesRouter);
       _defaultRouteInformationProvider?.dispose();
       _defaultRouteInformationProvider = null;
-      if (oldWidget == null || widget.navigatorKey != oldWidget.navigatorKey) {
+      if (_navigator == null || widget.navigatorKey != oldWidget!.navigatorKey) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/app.dart:1173]: This change doesn't seem to be part of "fix a MaterialApp NNBD issue" and there's no test for it. Maybe you didn't intend to include it?

**Review by HansMuller — COMMENTED** [review]
Nice fix; just one comment

File: packages/flutter/test/material/material_test.dart [hunk 2]
```diff
@@ -64,6 +64,20 @@ class ElevationColor {
...
+  // Regression test for https://github.com/flutter/flutter/issues/81504
+  testWidgets('MaterialApp.home nullable and update test', (WidgetTester tester) async {
+    // _WidgetsAppState._usesNavigator == true
+    await tester.pumpWidget(const MaterialApp(home: SizedBox.shrink()));
+
+    // _WidgetsAppState._usesNavigator == false
+    await tester.pumpWidget(const MaterialApp()); // Do not crash!
+
+    // _WidgetsAppState._usesNavigator == true
+    await tester.pumpWidget(const MaterialApp(home: SizedBox.shrink())); // Do not crash!
```
Comment (xu-baolin) [inline packages/flutter/test/material/material_test.dart:76]: Actually,this line test the change of the patch that you pointed out.
We should initial the '_navigator' if it equals null during update widget.

Comment (xu-baolin) [reply packages/flutter/test/material/material_test.dart:76]: I also describe it at the description:"Also, fix a crashing when update MaterialApp.home to non-null from null."

Comment (HansMuller) [reply packages/flutter/test/material/material_test.dart:76]: Got it, thanks for the explanation.

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by HansMuller** [comment]
Google testing failure appears to be unrelated to this PR. Confirming now.

**Comment by HansMuller** [comment]
Verified that the Google testing failure was a fluke. The failing test succeeded upon (deflake) running again ([internal google link](https://test.corp.google.com/ui#id=OCL:372387426:BASE:372388516:1620328218060:1f8d6fd9)).

**Comment by xu-baolin** [comment]
Add auto-merge label again.

**Pull request merged**